// plivo/dist/rest/client.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Client = undefined;

var _request = require('./request.js');

var _utils = require('./utils');

var _package = require('../../package.json');

var _call = require('../resources/call');

var _accounts = require('../resources/accounts');

var _applications = require('../resources/applications');

var _conferences = require('../resources/conferences');

var _endpoints = require('../resources/endpoints');

var _messages = require('../resources/messages');

var _numbers = require('../resources/numbers');

var _pricings = require('../resources/pricings');

var _recordings = require('../resources/recordings');

var _plivoxml = require('../utils/plivoxml');

var _security = require('../utils/security');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

exports.Response = function () {
  return new _plivoxml.Response();
};

exports.validateSignature = function (uri, nonce, signature, auth_token) {
  return (0, _security.validateSignature)(uri, nonce, signature, auth_token);
};

/**
 * Plivo API client which can be used to access the Plivo APIs.
 * To set a proxy or timeout, pass in options.proxy (url) or options.timeout (number in ms)
 * You can also pass in additional parameters accepted by the node requests module.
 */

var Client = exports.Client = function Client(authId, authToken, options) {
  _classCallCheck(this, Client);

  if (!(this instanceof Client)) {
    return new Client(authId, authToken, options);
  }
  authId = authId || process.env.PLIVO_AUTH_ID;
  authToken = authToken || process.env.PLIVO_AUTH_TOKEN;

  if (authId == null) {
    throw new Error('Please provide authId');
  }
  if (authToken == null) {
    throw new Error('Please provide authToken');
  }

  options = Object.assign({}, {
    authId: authId,
    authToken: authToken,
    version: 'v1',
    url: 'https://api.plivo.com/v1/Account/' + authId,
    userAgent: 'plivo-node' + '/' + (_package.version || 'Unknown Version') + ' (Node: ' + process.version + ')'
  }, options);

  var client = (0, _utils.camelCaseRequestWrapper)((0, _request.Request)(options));

  this.calls = new _call.CallInterface(client);
  this.accounts = new _accounts.AccountInterface(client);
  this.subaccounts = this.subAccounts = new _accounts.SubaccountInterface(client);
  this.applications = new _applications.ApplicationInterface(client);
  this.conferences = new _conferences.ConferenceInterface(client);
  this.endpoints = new _endpoints.EndpointInterface(client);
  this.messages = new _messages.MessageInterface(client);
  this.numbers = new _numbers.NumberInterface(client);
  this.pricings = new _pricings.PricingInterface(client);
  this.recordings = new _recordings.RecordingInterface(client);
};
